### PR TITLE
Feature/add discount price

### PR DESCRIPTION
**Pull request by SungjunApp**
ticket: https://pixlee.atlassian.net/browse/EN-3760

### Background
Since [Gallery Badge is on Phrase 1](https://pixlee-wiki.atlassian.net/wiki/spaces/PROD/pages/1382809646/Gallery+Badges#Beta-Customers), Mobile SDKs need to accommodate its data and UI.

### a new example to the demo app
- how to access the example: Launch the app > main screen > tap `Product View` button

### [Data] Added new fields to PXLProduct
- these fields are added to provide discount price information
```java
#!java
public class PXLProduct implements Parcelable {
...
    @FieldDate
    @Json(name = "sales_start_date")
    public Date salesStartDate;

    @FieldDate
    @Json(name = "sales_end_date")
    public Date salesEndDate;

    @FieldBigDecimal
    @Json(name = "sales_price")
    public BigDecimal salesPrice;
...
}
```

### [UI] Added a new option to PXLPhotoProductView for sales price
FYI, the color and size of the sales price and the default price are fully customizable via CurrencyTextStyle class.
- discountPriceTextStyle: ProductViewHolder.DiscountPrice(..) is added to PXLPhotoProductView.Configuration
  - priceTextStyle = CurrencyTextStyle
    - isCurrencyLeading:Boolean is added to CurrencyTextStyle. priceTextStyle has the same change as well.
      - true: display currency symbol in front of the price
      - false: display currency symbol after the price

  - DiscountLayout's options

|.CROSS_THROUGH|.WAS_OLD_PRICE|.WITH_DISCOUNT_LABEL|
|---------|---------|---------|
||||

|No Option selected or no sales price|Demo|
|---------|---------|
||

- Code Example
  - To display sales price, you need to set `discountPriceTextStyle = ProductViewHolder.DiscountPrice(..)`. Please see the example codes below.
  ```kotlin
  #!kotlin
  pxlPhotoProductView
            .setContent(photoInfo = ...,
                headerConfiguration = ...,
                configuration = ProductViewHolder.Configuration().apply {
                    circleIcon = ...
                    mainTextStyle = ...
                    subTextStyle = ..
                    bookmarkDrawable = ..
                    priceTextStyle = ...
                    discountPriceTextStyle = ProductViewHolder.DiscountPrice(
                            discountLayout = ProductViewHolder.DiscountLayout.WITH_DISCOUNT_LABEL,
                            priceTextStyle = CurrencyTextStyle().apply {
                                isCurrencyLeading = true
                                defaultCurrency = "EUR" // or null
                                leftText = TextStyle().apply {
                                    color = Color.RED
                                    size = 24.px
                                    sizeUnit = TypedValue.COMPLEX_UNIT_PX
                                    typeface = null
                                }

                                rightText = TextStyle().apply {
                                    color = Color.RED
                                    size = 14.px
                                    sizeUnit = TypedValue.COMPLEX_UNIT_PX
                                    typeface = null
                                }
                            }
                    )
                },
                bookmarkMap = ..,
                onProductClicked = {
                    ...
                })
  ```

### CODE REVIEW

File: app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+    fun makeProducts(discountLayout: ProductViewHolder.DiscountLayout?, products: List<PXLProduct>, showBookmarkIcon: Boolean = false, showShopIcon: Boolean = false): ProductAdapter {
+        return ProductAdapter(
+                configuration = configuration.copy().apply {
+                    if (discountLayout != null)
+                        discountPriceTextStyle = ProductViewHolder.DiscountPrice(
+                                discountLayout = discountLayout,
+                                priceTextStyle = CurrencyTextStyle().apply {
+                                    isCurrencyLeading = true
+                                    defaultCurrency = "EUR" // or null
```
Comment (jchen123) [inline app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt:161]: should this be USD?

File: app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+            rvOption3All.adapter = makeProducts(ProductViewHolder.DiscountLayout.WITH_DISCOUNT_LABEL, products, showBookmarkIcon = true, showShopIcon = true)
+            rvOption4All.adapter = makeProducts(null, products, showBookmarkIcon = true, showShopIcon = true)
+        }
+    }
+
+    val album by lazy {
+        PXLKtxAlbum(basicDS, analyticsDS).apply {
+            params = PXLKtxBaseAlbum.Params(
+                    // album images
+                    perPage = 15,
```
Comment (jchen123) [inline app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt:98]: is this adjustible?

File: pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/viewholder/ProductViewHolder.kt [hunk 3]
```diff
@@ -55,73 +74,178 @@ class ProductViewHolder(override val containerView: View) : RecyclerView.ViewHol
...
+                val discountPercentage = product.getDiscountPercentage()
+                offLable = if (discountLayout == DiscountLayout.WITH_DISCOUNT_LABEL && discountPercentage != null) {
+                    " ${tvSub.context.getString(R.string.percent_off, "$discountPercentage")}"
+                } else {
+                    ""
+                }
+                "$salesIntegerPrice$salesDecimalPrice $defaultPrinceString$offLable"
             } else {
-                price.getFractionalPart()
+                defaultPrinceString
```
Comment (jchen123) [inline pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/viewholder/ProductViewHolder.kt:144]: price, not prince right?

File: app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+                ?: ""
+    }
+
+    val mockWebServer: MockWebServer by lazy {
+        MockWebServer()
+    }
+
+    lateinit var basicDS: KtxBasicDataSource
+    lateinit var analyticsDS: KtxAnalyticsDataSource
+    fun setupMockedWebServer() {
```
Comment (jharrilim) [inline app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt:270]: Is this code for tests or demos? Not sure if it is worth the hassle, but if it is you might prefer to put this in an extension function in a different file, and only conditionally compile that file using `sourceSets.main.java.exclude "filename.kt"`.

**Review by jharrilim — APPROVED** [review]
neat! smart use of `with/let/apply/also`

File: app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt (hunk 2 above)
Comment (SungjunApp) [reply app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt:98]: yes! they are all adjustable

- val searchId: SearchId,
  ```kotlin
   sealed class SearchId {
        class Album(val id: String) : SearchId()
        class Product(val sku: String) : SearchId()
    }
  ``` 
- var perPage: Int = 30,
- var filterOptions: PXLAlbumFilterOptions? = null,
- var sortOptions: PXLAlbumSortOptions? = null

File: app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt (hunk 4 above)
Comment (SungjunApp) [reply app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt:270]: - all codes in /app directory are for the demo app. 
- This is for the demo app but it mocks the server and emmit the response body from a JSON file stored in the project. It's worth moving the unnecessary codes to another wrapper class. I'll move the mock server things to a new class. Thanks @jharrilim

File: app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt (hunk 1 above)
Comment (SungjunApp) [reply app/src/main/java/com/pixlee/pixleeandroidsdk/ui/uioptions/ProductViewFragment.kt:153]: That's the demonstration part of the demo. 
- app/ is the demo app. ProductViewFragment.kt is located under app/...../ directory.

This code is to show the SDK users how the SDK can be used. The SDK users can copy and paste this code into their project and change it for their needs

**Review by jchen123 — APPROVED** [review]

**Pull request merged**